src/action/ActionImpl.ts

```diff
@@ -0,0 +1,74 @@
+import { ReactElement, JSXElementConstructor, ReactNode } from "react";
+import type { BaseAction, Action } from "../types";
+
+interface ActionImplOptions {
+  parent?: ActionImpl;
+}
+
+export class ActionImpl implements Action {
```

Comment:
Users define actions by passing in an object like so:

```
{
  id: 123,
  name: "Post 1",
  parent: "parentId"
}
```

and we transform that into our internal `ActionImpl` object, which will maintain `parent`, `children`, and `ancestor` relationships outside of React land.